File: services/event-streaming-service/src/runtime/flink/job_graph.rs
//! HTTP proxy for Flink JobManager job-graph endpoint.
//!
//! Backs the REST handler `GET /topologies/:id/job-graph`. Surfaces the
//! Flink-native vertex/edge graph (the same payload the Flink Web UI
//! uses) so the front-end can render it with cytoscape.

use serde_json::Value;

use crate::runtime::flink::FlinkRuntimeConfig;

#[derive(Debug, thiserror::Error)]
pub enum JobGraphError {
    #[error("missing flink_deployment_name on topology")]
    NotDeployed,
    #[error("flink jobmanager returned status {0}")]
    Status(u16),
    #[error("http: {0}")]
    Http(#[from] reqwest::Error),
    #[error("invalid response: {0}")]
    Body(String),
}

/// Fetch and normalise the job graph. Output shape:
///
/// ```json
/// {
///   "job_id": "...",
///   "vertices": [{ "id": "...", "name": "...", "parallelism": 4 }],
///   "edges":    [{ "source": "...", "target": "..." }],
///   "raw":      { ... }   // verbatim Flink response for debugging
/// }
/// ```
pub async fn fetch_job_graph(
    cfg: &FlinkRuntimeConfig,
    deployment: &str,
    namespace: &str,
    job_id: Option<&str>,
) -> Result<Value, JobGraphError> {
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .build()?;
    let base = cfg.jobmanager_url(deployment, namespace);
    let resolved_job = match job_id {
        Some(id) => id.to_string(),
        None => discover(&client, &base).await?,
    };
    let url = format!("{base}/jobs/{resolved_job}");
    let resp = client.get(&url).send().await?;
    if !resp.status().is_success() {
        return Err(JobGraphError::Status(resp.status().as_u16()));
    }
    let raw: Value = resp.json().await?;
    Ok(normalise(raw, &resolved_job))
}

async fn discover(client: &reqwest::Client, base: &str) -> Result<String, JobGraphError> {
    let body: Value = client
        .get(&format!("{base}/jobs"))
        .send()
        .await?
        .error_for_status()
        .map_err(|e| JobGraphError::Status(e.status().map(|s| s.as_u16()).unwrap_or(0)))?
        .json()
        .await?;
    let jobs = body
        .get("jobs")
        .and_then(|v| v.as_array())
        .ok_or_else(|| JobGraphError::Body("no 'jobs' array".into()))?;
    let chosen = jobs
        .iter()
        .find(|j| j.get("status").and_then(|s| s.as_str()) == Some("RUNNING"))
        .or_else(|| jobs.first())
        .and_then(|j| j.get("id").and_then(|s| s.as_str()))
        .ok_or_else(|| JobGraphError::Body("no jobs reported".into()))?
        .to_string();
    Ok(chosen)
}

/// Pure helper: pluck the cytoscape-friendly subset out of a Flink
/// `/jobs/{id}` response.
pub fn normalise(raw: Value, job_id: &str) -> Value {
    let vertices: Vec<Value> = raw
        .get("vertices")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .map(|v| {
                    serde_json::json!({
                        "id": v.get("id").cloned().unwrap_or(Value::Null),
                        "name": v.get("name").cloned().unwrap_or(Value::Null),
                        "parallelism": v.get("parallelism").cloned().unwrap_or(Value::Null),
                        "status": v.get("status").cloned().unwrap_or(Value::Null),
                    })
                })
                .collect()
        })
        .unwrap_or_default();
    let edges: Vec<Value> = raw
        .get("plan")
        .and_then(|p| p.get("nodes"))
        .and_then(|n| n.as_array())
        .map(|nodes| {
            let mut out = Vec::new();
            for node in nodes {
                if let (Some(target), Some(inputs)) = (
                    node.get("id").and_then(|v| v.as_str()),
                    node.get("inputs").and_then(|v| v.as_array()),
                ) {
                    for input in inputs {
                        if let Some(src) = input.get("id").and_then(|v| v.as_str()) {
                            out.push(serde_json::json!({
                                "source": src,
                                "target": target,
                            }));
                        }
                    }
                }
            }
            out
        })
        .unwrap_or_default();
    serde_json::json!({
        "job_id": job_id,
        "vertices": vertices,
        "edges": edges,
        "raw": raw,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn normalise_extracts_vertices_and_edges_from_flink_payload() {
        let payload = json!({
            "vertices": [
                {"id": "v1", "name": "Source", "parallelism": 2, "status": "RUNNING"},
                {"id": "v2", "name": "Sink",   "parallelism": 4, "status": "RUNNING"},
            ],
            "plan": {
                "nodes": [
                    {"id": "v2", "inputs": [{"id": "v1"}]}
                ]
            }
        });
        let out = normalise(payload, "job-x");
        assert_eq!(out["job_id"], "job-x");
        assert_eq!(out["vertices"].as_array().unwrap().len(), 2);
        assert_eq!(out["edges"][0]["source"], "v1");
        assert_eq!(out["edges"][0]["target"], "v2");
    }
}

File: services/ingestion-replication-service/internal/runtime/flink/job_graph.go
package flink

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"time"
)

// JobGraphErrorKind mirrors the variants of JobGraphError in the Rust source.
type JobGraphErrorKind int

const (
	JobGraphErrUnknown JobGraphErrorKind = iota
	JobGraphErrNotDeployed
	JobGraphErrStatus
	JobGraphErrHTTP
	JobGraphErrBody
)

// JobGraphError mirrors event_streaming::runtime::flink::job_graph::JobGraphError.
//
// The Kind field discriminates between the Rust enum variants; Status
// carries the HTTP status code for `Status` errors; Cause carries the
// wrapped HTTP error for `HTTP` errors.
type JobGraphError struct {
	Kind    JobGraphErrorKind
	Status  int
	Message string
	Cause   error
}

func (e *JobGraphError) Error() string {
	switch e.Kind {
	case JobGraphErrNotDeployed:
		return "missing flink_deployment_name on topology"
	case JobGraphErrStatus:
		return fmt.Sprintf("flink jobmanager returned status %d", e.Status)
	case JobGraphErrHTTP:
		return fmt.Sprintf("http: %v", e.Cause)
	case JobGraphErrBody:
		return fmt.Sprintf("invalid response: %s", e.Message)
	default:
		if e.Message != "" {
			return e.Message
		}
		return "unknown job graph error"
	}
}

func (e *JobGraphError) Unwrap() error { return e.Cause }

// ErrJobGraphNotDeployed is the sentinel for the `NotDeployed` variant.
// Returned by the handler layer, kept here so callers can switch on it.
var ErrJobGraphNotDeployed = &JobGraphError{Kind: JobGraphErrNotDeployed}

// JobGraphVertex mirrors the cytoscape-friendly vertex projection
// produced by `normalise`.
type JobGraphVertex struct {
	ID          any `json:"id"`
	Name        any `json:"name"`
	Parallelism any `json:"parallelism"`
	Status      any `json:"status"`
}

// JobGraphEdge mirrors the cytoscape-friendly edge projection produced
// by `normalise`. Edges are derived from the Flink JSON `plan.nodes[].inputs`
// adjacency list.
type JobGraphEdge struct {
	Source string `json:"source"`
	Target string `json:"target"`
}

// JobGraph is the normalised front-end-friendly job graph.
//
// Equivalent to the JSON object the Rust `normalise` returns:
//
//	{ "job_id": "...", "vertices": [...], "edges": [...], "raw": {...} }
//
// Raw is the verbatim Flink response, preserved for debugging.
type JobGraph struct {
	JobID    string                 `json:"job_id"`
	Vertices []JobGraphVertex       `json:"vertices"`
	Edges    []JobGraphEdge         `json:"edges"`
	Raw      map[string]any         `json:"raw"`
}

// HTTPDoer is the minimal HTTP client surface FetchJobGraph needs. Lets
// tests inject a fake without spinning up a real server.
type HTTPDoer interface {
	Do(req *http.Request) (*http.Response, error)
}

// DefaultHTTPClient is the production client used when callers pass nil.
// 10 s timeout matches the Rust `reqwest` builder.
func DefaultHTTPClient() *http.Client {
	return &http.Client{Timeout: 10 * time.Second}
}

// FetchJobGraph ports event_streaming::runtime::flink::job_graph::fetch_job_graph.
//
// It hits the JobManager REST API and returns the normalised cytoscape
// payload. When jobID is empty the function discovers the running job
// off `/jobs` (preferring `RUNNING`, falling back to the first entry).
func FetchJobGraph(ctx context.Context, client HTTPDoer, cfg FlinkRuntimeConfig, deployment, namespace, jobID string) (*JobGraph, error) {
	if client == nil {
		client = DefaultHTTPClient()
	}
	base := cfg.JobManagerURL(deployment, namespace)
	resolved := jobID
	if resolved == "" {
		discovered, err := discoverJob(ctx, client, base)
		if err != nil {
			return nil, err
		}
		resolved = discovered
	}
	url := base + "/jobs/" + resolved
	raw, err := getJSON(ctx, client, url)
	if err != nil {
		return nil, err
	}
	return Normalise(raw, resolved), nil
}

func discoverJob(ctx context.Context, client HTTPDoer, base string) (string, error) {
	body, err := getJSON(ctx, client, base+"/jobs")
	if err != nil {
		return "", err
	}
	rawJobs, ok := body["jobs"].([]any)
	if !ok {
		return "", &JobGraphError{Kind: JobGraphErrBody, Message: "no 'jobs' array"}
	}
	// Prefer the first RUNNING job, falling back to the first entry —
	// matches the Rust source's `find().or_else(first())` chain.
	var chosen map[string]any
	for _, entry := range rawJobs {
		obj, ok := entry.(map[string]any)
		if !ok {
			continue
		}
		if status, _ := obj["status"].(string); status == "RUNNING" {
			chosen = obj
			break
		}
	}
	if chosen == nil && len(rawJobs) > 0 {
		if obj, ok := rawJobs[0].(map[string]any); ok {
			chosen = obj
		}
	}
	if chosen == nil {
		return "", &JobGraphError{Kind: JobGraphErrBody, Message: "no jobs reported"}
	}
	id, ok := chosen["id"].(string)
	if !ok || id == "" {
		return "", &JobGraphError{Kind: JobGraphErrBody, Message: "no jobs reported"}
	}
	return id, nil
}

// Normalise plucks the cytoscape-friendly subset out of a Flink
// `/jobs/{id}` response. Pure, no I/O — exposed so callers (and tests)
// can reuse it without hitting the network.
func Normalise(raw map[string]any, jobID string) *JobGraph {
	vertices := make([]JobGraphVertex, 0)
	if rawVerts, ok := raw["vertices"].([]any); ok {
		for _, v := range rawVerts {
			obj, ok := v.(map[string]any)
			if !ok {
				continue
			}
			vertices = append(vertices, JobGraphVertex{
				ID:          obj["id"],
				Name:        obj["name"],
				Parallelism: obj["parallelism"],
				Status:      obj["status"],
			})
		}
	}
	edges := make([]JobGraphEdge, 0)
	if plan, ok := raw["plan"].(map[string]any); ok {
		if nodes, ok := plan["nodes"].([]any); ok {
			for _, n := range nodes {
				node, ok := n.(map[string]any)
				if !ok {
					continue
				}
				target, ok := node["id"].(string)
				if !ok {
					continue
				}
				inputs, ok := node["inputs"].([]any)
				if !ok {
					continue
				}
				for _, in := range inputs {
					inputObj, ok := in.(map[string]any)
					if !ok {
						continue
					}
					if src, ok := inputObj["id"].(string); ok {
						edges = append(edges, JobGraphEdge{Source: src, Target: target})
					}
				}
			}
		}
	}
	return &JobGraph{
		JobID:    jobID,
		Vertices: vertices,
		Edges:    edges,
		Raw:      raw,
	}
}

func getJSON(ctx context.Context, client HTTPDoer, url string) (map[string]any, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, &JobGraphError{Kind: JobGraphErrHTTP, Cause: err}
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, &JobGraphError{Kind: JobGraphErrHTTP, Cause: err}
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, &JobGraphError{Kind: JobGraphErrStatus, Status: resp.StatusCode}
	}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, &JobGraphError{Kind: JobGraphErrHTTP, Cause: err}
	}
	var obj map[string]any
	if err := json.Unmarshal(body, &obj); err != nil {
		return nil, &JobGraphError{Kind: JobGraphErrBody, Message: err.Error()}
	}
	return obj, nil
}

// IsNotDeployed reports whether err is the NotDeployed sentinel — used
// by the REST handler to map to a 409.
func IsNotDeployed(err error) bool {
	var jg *JobGraphError
	if errors.As(err, &jg) {
		return jg.Kind == JobGraphErrNotDeployed
	}
	return false
}
